src/gestalts/types.ts

```diff
@@ -22,13 +23,18 @@ type GestaltIdentityKey = Opaque<'GestaltIdentityKey', string>;
 
 type GestaltKeySet = Record<GestaltKey, null>;
 type GestaltMatrix = Record<GestaltKey, GestaltKeySet>;
-type GestaltNodes = Record<GestaltNodeKey, NodeInfo>;
+type GestaltNodes = Record<GestaltNodeKey, GestaltNodeInfo>;
 type GestaltIdentities = Record<GestaltIdentityKey, IdentityInfo>;
 type Gestalt = {
   matrix: GestaltMatrix;
   nodes: GestaltNodes;
   identities: GestaltIdentities;
 };
+// Need a type that plays nice with the DB. Used internally only.
+type GestaltNodeInfo = {
+  id: NodeIdEncoded;
+  chain: ChainData;
+};
```

Comment:
The `GestaltNodeInfo` is new. What was the `NodeInfo` before, and why was this moved into `gestalts/types.ts`?

Reply:
This is just for use in the DB, NodeInfo is used elsewhere but due to a problem with putting and getting it from the DB i needed a version I could use internally in gestalts.

Reply:
Ok I reproduced this.

The `NodeInfo` is using `NodeId`, and `Uint8Array` requires marshalling when going to and from JSON, because the DB will encode the data prior.

```
// '{"b":{"0":97,"1":98,"2":99}}'
s = JSON.stringify({ b: Uint8Array.from([97, 98, 99]) });
// { b: { '0': 97, '1': 98, '2': 99 } }
JSON.parse(s);
```

Note that this problem has been addressed in Node's buffer: https://nodejs.org/api/buffer.html#buftojson. The native json stringify will automatically encode the buffer into a special representation, however during JSON.parse, you have to then unmarshal it manually back to a buffer object.

The `JSON.parse` has a second parameter called `reviver` that can be used to unmarshal JSON strings in a custom way, and convert these back into ID. Same with `JSON.stringify` called `replacer` which can be used to marshal IDs appropriately.

So the choices are:

1. Use `NodeIdEncoded` when storing these `NodeId` into DB as values (as keys they can be stored as raw buffers as designed).
2. Create custom marshalling/unmarshalling to deal with this.

I think we should add a `toJSON` for `IdInternal` class, since we can use this: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#tojson_behavior.

At the very least this allows us to potentially decode from a custom JSON encoding as well. This seems to have general implications across the code wherever we use `NodeId` as values in the DB.

@joshuakarp thoughts?

Reply:
Remember that `NodeInfo` is essentially a serialised map of sigchain data for some particular node, and this is stored within the `GestaltGraph` (found during discovery, etc).

I'd expect that the `GestaltGraph` is eventually going to be displayed in a human readable fashion. Therefore, the marshalling back and forth from the `UInt8Array` `NodeId` seems to be unnecessary, as in the end we'd want to display the encoded version of the `NodeId` anyway.

Would you agree with the above @CMCDragonkai ?

Reply:
I'm going to add `toJSON` method into the `IdInternal`. This should return something like:

```ts
{type: "IdInternal", data: [97, 98, 99]}
```

This will make it similar to Node buffers in this case, this is at least one step.

In addition, I'll add the `equals` method to allow easier more efficient equality comparisons between `Id` objects, and overloading `==` is not possible atm unless one is a primitive first.

Reply:
@joshuakarp currently `NodeInfo` is using `NodeId` and not `NodeIdEncoded`. If the `NodeInfo` will end up always showing the encoded version, then `NodeInfo` should still be used, but have its `id` become `NodeIdEncoded`, like:

```
type NodeInfo = {
  id: NodeIdEncoded;
  chain: ChainData;
};
```

Right now when I use `ag` to find all references of `NodeInfo`, I see it used in `GestaltGraph`, `NodeConnection` and `Discovery`. It seems `Discovery` and `GestaltGraph` are both limited to using the encoded node id for now. And `NodeConnection` is only using it as a comment.

In that case, for now it makes sense to continue using `NodeInfo`, but just update the type.

We should then understand that `NodeInfo` represents a human readable POJO of node information...

We may need to consider how `NodeInfo` would be used in the future to see if it is worth it to keep it in its raw buffer.

Reply:
Using the Reviver and  Replacer seems like a good solution here. but we will have to add the `toJSON` method to `IdInternal` in `js-id`.

Reply:
> In that case, for now it makes sense to continue using NodeInfo, but just update the type.
> 
> We should then understand that NodeInfo represents a human readable POJO of node information...

I agree with this, and so agree that `NodeIdEncoded` is the correct type to be using here.

Reply:
@tegefaulkes can we switch back to `NodeInfo` in that case and just change to `NodeIdEncoded` atm? I just checked `IdentityInfo`, and that also uses string forms of `ProviderId` and `IdentityId` as well.

Seems like `NodeInfo` is primarily used by gestalts and sigchain related work, and we will refactor this when we get there if need be.

Reply:
Did some quick testing.
```ts
// Added to IdInternal
  public toJSON(): string {
    return this.toString();
  }
```

example of usage.
```ts
const id = IdInternal.fromString('testString');
    const testObj = {
      test: 'test',
      id
    }

    const jsonId = JSON.stringify(id)
    const jsonObj = JSON.stringify(testObj);

    const revivedId = JSON.parse(jsonId, (_, value) => IdInternal.fromString(value));
    const revivedObj = JSON.parse(jsonObj, (key, value) => {
      if (key === 'id') return IdInternal.fromString(value)
      return value
    });

    expect(revivedId).toEqual(id);
    expect(revivedObj).toEqual(testObj);
  })
```

So it seems to work fine, but the only annoyance is that we have to tailor the reviver for each use.

Reply:
I'll change it back to nodeInfo and switch that to using NodeIdEncoded. 

ignore my two comments, they're out of place in this discussion since the comments here didn't update while I was writing them.

Reply:
Going to resolve this for now then since we are using `NodeInfo` with `NodeIdEncoded`. But 3.3.0 allows for the json marshalling if we need it.